Python/Product/Analysis/DocumentBuffer.cs

```diff
@@ -0,0 +1,94 @@
+﻿// Python Tools for Visual Studio
+// Copyright(c) Microsoft Corporation
+// All rights reserved.
+//
+// Licensed under the Apache License, Version 2.0 (the License); you may not use
+// this file except in compliance with the License. You may obtain a copy of the
+// License at http://www.apache.org/licenses/LICENSE-2.0
+//
+// THIS CODE IS PROVIDED ON AN  *AS IS* BASIS, WITHOUT WARRANTIES OR CONDITIONS
+// OF ANY KIND, EITHER EXPRESS OR IMPLIED, INCLUDING WITHOUT LIMITATION ANY
+// IMPLIED WARRANTIES OR CONDITIONS OF TITLE, FITNESS FOR A PARTICULAR PURPOSE,
+// MERCHANTABLITY OR NON-INFRINGEMENT.
+//
+// See the Apache Version 2.0 License for specific language governing
+// permissions and limitations under the License.
+
+using System;
+using System.Collections.Generic;
+using System.Linq;
+using System.Text;
+using Microsoft.PythonTools.Parsing;
+
+namespace Microsoft.PythonTools.Analysis {
+    class DocumentBuffer {
+        public int Version { get; private set; } = -1;
+        public StringBuilder Text { get; } = new StringBuilder();
+
+        public void Reset(int version, string content) {
+            Version = version;
+            Text.Clear();
+            Text.Append(content);
+        }
+
+        public void Update(IEnumerable<DocumentChangeSet> changes) {
+            foreach (var change in changes) {
+                Update(change);
```

Comment:
If one of the updates fails with exception, the others won't be processed. Is that ok?

Reply:
Yep. We do need to add an "out of sync" message for the server to raise in this case though (and some others), as there's no way to return failure from a document changed event.